wal/src/message_queue_impl/region.rs

```diff
@@ -0,0 +1,761 @@
+                region_id: self.region_id,
+            })?;
+
+        info!("Finished to read all logs from region, region id:{}, log topic:{}, terminate offset:{}", self.region_id, 
+            self.log_topic, current_high_watermark);
+
+        Ok(MessageQueueLogIterator::new(
+            self.region_id,
+            Some(current_high_watermark),
+            consume_iter,
+            self.log_encoding.clone(),
+        ))
+    }
+
+    /// Mark the delete to sequence number of table.
+    async fn mark_delete_up_to(
+        &self,
+        table_id: TableId,
+        sequence_num: SequenceNumber,
+    ) -> Result<()> {
+        debug!(
+            "Mark deleted entries up to sequence_num:{}, region_id:{}",
+            sequence_num, self.region_id
+        );
+
+        self.region_meta
+            .mark_table_deleted(table_id, sequence_num)
+            .await
+            .context(MarkDeleted {
+                region_id: self.region_id,
+                table_id,
+            })
+    }
+
+    /// Get meta data by table id.
+    async fn get_table_meta(&self, table_id: TableId) -> Result<TableMetaData> {
+        self.region_meta
+            .get_table_meta_data(table_id)
+            .await
+            .context(GetTableMeta {
+                region_id: self.region_id,
+                table_id,
+            })
+    }
+
+    /// Get meta data snapshot of whole region.
+    ///
+    /// NOTICE: should freeze whole region before calling.
+    async fn make_meta_snapshot(&self) -> RegionMetaSnapshot {
+        self.region_meta.make_snapshot().await
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug)]
+pub struct MessageQueueLogIterator<C: ConsumeIterator> {
+    /// Id of region
+    region_id: RegionId,
+
+    /// Polling's end point
+    ///
+    /// While fetching in slave node, it will be set to `None`, and
+    /// reading will not stop.
+    /// Otherwise, it will be set to high watermark.
+    terminate_offset: Option<Offset>,
+
+    /// Terminated flag
+    is_terminated: bool,
+    /// Consume Iterator of message queue
+    iter: C,
+    /// Used to encode/decode the logs
+    log_encoding: CommonLogEncoding,
+    /// See the same problem in https://github.com/CeresDB/ceresdb/issues/120
+    previous_value: Vec<u8>,
+    // TODO: timeout
+}
+
+#[allow(unused)]
+impl<C: ConsumeIterator> MessageQueueLogIterator<C> {
+    fn new(
+        region_id: RegionId,
+        terminate_offset: Option<Offset>,
+        iter: C,
+        log_encoding: CommonLogEncoding,
+    ) -> Self {
+        Self {
+            region_id,
+            terminate_offset,
+            iter,
+            is_terminated: false,
+            log_encoding,
+            previous_value: Vec::new(),
+        }
+    }
+}
+
+#[allow(unused)]
+impl<C: ConsumeIterator> MessageQueueLogIterator<C> {
+    pub async fn next_log_entry(&mut self) -> Result<Option<LogEntry<&'_ [u8]>>> {
+        if self.is_terminated && self.terminate_offset.is_some() {
+            info!(
```

Comment:
I guess this log should be set debug level.

Reply:
It will just print once while finishing to poll, I think `info` may be ok?